File: doctor.py
"""Clink doctor: diagnose environment and configuration issues."""

from __future__ import annotations

import shutil
import sys

from config import CONFIG_FILE, load_config


def _check(label: str, ok: bool, hint: str = "") -> bool:
    mark = "✓" if ok else "✗"
    suffix = f"  → {hint}" if hint and not ok else ""
    print(f"  {mark} {label}{suffix}")
    return ok


def run_doctor() -> None:
    """Check environment, dependencies, and configuration."""
    print("\nClink Doctor\n")
    all_ok = True

    # Python version
    v = sys.version_info
    all_ok &= _check(
        f"Python {v.major}.{v.minor}.{v.micro}",
        v >= (3, 10),
        "need Python >= 3.10",
    )

    # Claude CLI
    claude_path = shutil.which("claude")
    all_ok &= _check(
        "Claude Code CLI",
        claude_path is not None,
        "npm i -g @anthropic-ai/claude-code",
    )

    # pip dependencies
    pkg_import_map = {
        "pyyaml": "yaml",
        "qq-botpy": "botpy",
        "claude-agent-sdk": "claude_agent_sdk",
    }
    req_file = __import__("pathlib").Path(__file__).parent / "requirements.txt"
    if req_file.exists():
        missing: list[str] = []
        for line in req_file.read_text().splitlines():
            pkg = line.strip()
            if pkg and not pkg.startswith("#"):
                import_name = pkg_import_map.get(pkg, pkg.replace("-", "_"))
                try:
                    __import__(import_name)
                except ImportError:
                    missing.append(pkg)
        all_ok &= _check(
            "pip dependencies",
            not missing,
            f"missing: {', '.join(missing)}. Run: pip install -r requirements.txt" if missing else "",
        )

    # Config file
    cfg_exists = CONFIG_FILE.exists()
    all_ok &= _check(
        f"Config file ({CONFIG_FILE})",
        cfg_exists,
        "run: python clink.py setup",
    )

    if cfg_exists:
        cfg = load_config()
        channel = cfg.get("channel", "")
        all_ok &= _check(
            f"Channel configured: {channel}",
            channel in ("terminal", "qq", "wecom"),
            "unknown channel",
        )

        # Check channel-specific config
        if channel == "qq":
            qq_cfg = cfg.get("qq", {})
            all_ok &= _check(
                "QQ Bot credentials",
                bool(qq_cfg.get("appid") and qq_cfg.get("secret")),
                "missing appid or secret",
            )
        elif channel == "wecom":
            wc = cfg.get("wecom", {})
            required_keys = ["corp_id", "corp_secret", "agent_id", "token", "encoding_aes_key"]
            missing_keys = [k for k in required_keys if not wc.get(k)]
            all_ok &= _check(
                "WeCom credentials",
                not missing_keys,
                f"missing: {', '.join(missing_keys)}" if missing_keys else "",
            )

    print()
    if all_ok:
        print("  All checks passed! Run: python clink.py start\n")
    else:
        print("  Some checks failed. Fix the issues above and re-run doctor.\n")

File: src/doctor.ts
import { execFileSync } from "node:child_process";
import { existsSync } from "node:fs";
import pc from "picocolors";
import { CONFIG_FILE } from "./config.js";
import { validateConfig, formatValidationIssues } from "./config-validate.js";

function which(cmd: string): string | null {
  try {
    return execFileSync("which", [cmd], { encoding: "utf-8" }).trim();
  } catch {
    return null;
  }
}

function check(label: string, ok: boolean, hint = ""): boolean {
  const mark = ok ? pc.green("✓") : pc.red("✗");
  const suffix = !ok && hint ? `  → ${hint}` : "";
  console.log(`  ${mark} ${label}${suffix}`);
  return ok;
}

export function runDoctor(): void {
  console.log("\nKlaus Doctor\n");
  let allOk = true;

  // Node.js version
  const [major] = process.versions.node.split(".").map(Number);
  allOk &&= check(
    `Node.js ${process.version}`,
    major >= 18,
    "need Node.js >= 18",
  );

  // Claude CLI
  const claudePath = which("claude");
  allOk &&= check(
    "Claude Code CLI",
    claudePath !== null,
    "npm i -g @anthropic-ai/claude-code",
  );

  // Config file existence
  const cfgExists = existsSync(CONFIG_FILE);
  allOk &&= check(
    `Config file (${CONFIG_FILE})`,
    cfgExists,
    "run: klaus setup",
  );

  // Config validation (reuse shared validation logic)
  if (cfgExists) {
    const result = validateConfig();
    if (result.valid) {
      const rawChannel = result.config.channel;
      const channels: string[] = Array.isArray(rawChannel)
        ? rawChannel.map(String)
        : [String(rawChannel)];
      allOk &&= check(`Config valid (channel: ${channels.join(", ")})`, true);

      // Check tunnel CLI prerequisites for web channel
      if (channels.includes("web")) {
        const webSection = (result.config.web as Record<string, unknown>) ?? {};
        const tunnel = webSection.tunnel;
        if (tunnel === true) {
          allOk &&= check(
            "cloudflared CLI (Quick Tunnel)",
            which("cloudflared") !== null,
            "brew install cloudflared",
          );
        } else if (typeof tunnel === "object" && tunnel !== null) {
          const provider = (tunnel as Record<string, unknown>).provider;
          if (provider === "cloudflare-quick" || provider === "cloudflare") {
            allOk &&= check(
              "cloudflared CLI",
              which("cloudflared") !== null,
              "brew install cloudflared",
            );
          } else if (provider === "ngrok") {
            allOk &&= check(
              "ngrok CLI",
              which("ngrok") !== null,
              "brew install ngrok",
            );
          }
        }
      }
    } else {
      allOk &&= check("Config validation", false, "see details below");
      console.log();
      console.log(formatValidationIssues(result.issues));
    }
  }

  console.log();
  if (allOk) {
    console.log(`  ${pc.green("All checks passed!")} Run: klaus start\n`);
  } else {
    console.log(
      "  Some checks failed. Fix the issues above and re-run doctor.\n",
    );
  }
}
